app/src/main/java/com/rikin/cinco/domain/GameViewModel.kt

```diff
@@ -184,26 +185,47 @@ class GameViewModel(private val clipboardHelper: ClipboardHelper) : ViewModel()
     }
 
     private fun updateKeyboard(row: RowState, keyboard: KeyboardState): KeyboardState {
-        val updatedRows = mutableListOf<KeyboardRowState>()
-        keyboard.keyRows.forEach { keyRow ->
-            val updatedKeys = mutableListOf<KeyState>()
-            keyRow.keys.forEach { key ->
-                // There is an issue here where a guess with a repeating letter in which the position
-                // of the first is incorrect, but the position of the second isn't can appear as
-                // incorrect. This is because we only look for the first letter in the row instead of
-                // all instances of a letter.
-                val tileForKey = row.tiles.find { it.letter == key.letter }
-                if (tileForKey != null && key.status.priority >= tileForKey.status.priority) {
-                    updatedKeys.add(key.copy(status = tileForKey.status))
-                } else {
-                    updatedKeys.add(key)
+
+        fun updateKeyMappings(): KeyboardState {
+            row.tiles.forEach { tile: TileState ->
+                val letter = tile.letter
+                val currentKey = keyboard.keyMappings[letter]
+                    ?: throw IllegalArgumentException("Could not find the letter $letter in key mappings.")
+                val tilePriority = tile.status.priority
+
+                if (tilePriority <= currentKey.status.priority) {
+                    keyboard.keyMappings[letter] = currentKey.copy(status = tile.status)
                 }
             }
-            updatedRows.add(KeyboardRowState(keys = updatedKeys))
+
+            return keyboard
         }
-        return KeyboardState(
-            keyRows = updatedRows
-        )
+
+        fun updateKeyboardRows(updatedState: KeyboardState): MutableList<KeyboardRowState> {
+            val updatedRows = mutableListOf<KeyboardRowState>()
+            keyboard.keyRows.forEach { keyRow ->
+                val updatedKeys = mutableListOf<KeyState>()
```

Comment:
We can use `map` here so that we don't need to create the `MutableList<KeyState>`

```kotlin
val updatedKeys = keyRow.keys.map { /* insert logic */ }
```